mlflow/projects/__init__.py

```diff
@@ -189,6 +194,29 @@ def _is_local_uri(uri):
     return not _GIT_URI_REGEX.match(uri)
 
 
+def _is_valid_version(repo, version):
+    """Verify the commit version of branch name is valid, raise exception if otherwise."""
+    from git.exc import GitCommandError
+    try:
+        return repo.git.rev_parse("--verify", version)
+    except GitCommandError:
+        # Second attempt to see if this is a branch
```

Comment:
This function is called in `_fetch_git_repo()`
```
    if version is not None and _is_valid_version(repo, version):
        repo.git.checkout(version)
```
Without this check, when an invalid `version` is provided, we get this:
```
=== Fetching project from https://github.com/adrian555/DocsDump#files/mlflow-projects/R into /var/folders/j6/bvpbjjrn3t1brs4dm9x04s5h0000gn/T/tmpggc02g4d ===
Traceback (most recent call last):
  File "/Users/wzhuang/anaconda3/envs/mlflow/bin/mlflow", line 11, in <module>
    sys.exit(cli())
  File "/Users/wzhuang/anaconda3/envs/mlflow/lib/python3.6/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/Users/wzhuang/anaconda3/envs/mlflow/lib/python3.6/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/Users/wzhuang/anaconda3/envs/mlflow/lib/python3.6/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/wzhuang/anaconda3/envs/mlflow/lib/python3.6/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/wzhuang/anaconda3/envs/mlflow/lib/python3.6/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/Users/wzhuang/anaconda3/envs/mlflow/lib/python3.6/site-packages/mlflow/cli.py", line 112, in run
    run_id=run_id,
  File "/Users/wzhuang/anaconda3/envs/mlflow/lib/python3.6/site-packages/mlflow/projects/__init__.py", line 135, in run
    storage_dir=storage_dir, block=block, run_id=run_id)
  File "/Users/wzhuang/anaconda3/envs/mlflow/lib/python3.6/site-packages/mlflow/projects/__init__.py", line 43, in _run
    git_username=git_username, git_password=git_password)
  File "/Users/wzhuang/anaconda3/envs/mlflow/lib/python3.6/site-packages/mlflow/projects/__init__.py", line 211, in _fetch_project
    _fetch_git_repo(parsed_uri, version, dst_dir, git_username, git_password)
  File "/Users/wzhuang/anaconda3/envs/mlflow/lib/python3.6/site-packages/mlflow/projects/__init__.py", line 242, in _fetch_git_repo
    repo.git.checkout(version)
  File "/Users/wzhuang/anaconda3/envs/mlflow/lib/python3.6/site-packages/git/cmd.py", line 548, in <lambda>
    return lambda *args, **kwargs: self._call_process(name, *args, **kwargs)
  File "/Users/wzhuang/anaconda3/envs/mlflow/lib/python3.6/site-packages/git/cmd.py", line 1014, in _call_process
    return self.execute(call, **exec_kwargs)
  File "/Users/wzhuang/anaconda3/envs/mlflow/lib/python3.6/site-packages/git/cmd.py", line 825, in execute
    raise GitCommandError(command, status, stderr_value, stdout_value)
git.exc.GitCommandError: Cmd('git') failed due to: exit code(1)
  cmdline: git checkout abc
  stderr: 'error: pathspec 'abc' did not match any file(s) known to git.'
```
Now, we will see this:
```
=== Fetching project from https://github.com/adrian555/DocsDump#files/mlflow-projects/R into /var/folders/j6/bvpbjjrn3t1brs4dm9x04s5h0000gn/T/tmpye9v0bgl ===
=== Got exception when validating git version 'abc' - please ensure that the version exists in the repo. Error: Cmd('git') failed due to: exit code(128)
  cmdline: git rev-parse --verify refs/remotes/origin/abc
  stderr: 'fatal: Needed a single revision' 
```
This looks like cleaner without the whole stack trace. Alternatively, of course, we can just wrap the try-catch on the checkout call.

Reply:
Gotcha - sorry, my original question was unclear - I meant to ask: in what scenario is it helpful to try to check that `version` is a branch the second time (using `repo.git.rev_parse("--verify", "refs/remotes/origin/%s" % version)`)?

Reply:
Basically, my preference would be to try to detect the branch only once since it keeps the method simpler, unless there's a common case that's handled via the `repo.git.rev_parse("--verify", "refs/remotes/origin/%s" % version))` logic

Reply:
Since `_is_valid_version()` is called before the `git.checkout()`, the `git rev-parse --verify branch_name` will fail even it is a valid branch, hence the second attempt to verify that the branch exists.

Regardless, this function has been removed.